Source/Foundation/bsfUtility/Utility/BsDynLibManager.h

```diff
@@ -5,37 +5,70 @@
 #include "Prerequisites/BsPrerequisitesUtil.h"
 #include "Utility/BsModule.h"
 
-namespace bs 
+#include "Utility/BsDynLib.h"
+
+namespace bs
 {
 	/** @addtogroup General
 	 *  @{
 	 */
 
-    /**
-	 * This manager keeps a track of all the open dynamic-loading libraries, opens them and returns references to 
+	/**
+	 * This manager keeps a track of all the open dynamic-loading libraries, opens them and returns references to
 	 * already-open libraries.
-	 *			
+	 *
 	 * @note	Not thread safe.
-     */
-    class BS_UTILITY_EXPORT DynLibManager : public Module<DynLibManager>
-    {
-    public:
-        DynLibManager() = default;
-        virtual ~DynLibManager();
-
-        /**
-         * Loads the given file as a dynamic library.
-         *
-         * @param[in]	name	The name of the library. The extension can be omitted.
-         */
-        DynLib* load(const String& name);
+	 */
+	class BS_UTILITY_EXPORT DynLibManager : public Module<DynLibManager>
+	{
+	public:
+		DynLibManager() = default;
+		virtual ~DynLibManager() = default;
+
+		/**
+		 * Loads the given file as a dynamic library.
+		 *
+		 * @param[in]	name	The name of the library. The extension can be omitted.
+		 */
+		DynLib* load(String name);
 
 		/** Unloads the given library. */
 		void unload(DynLib* lib);
 
 	protected:
-		Map<String, DynLib*> mLoadedLibraries;
-    };
+		struct DynLibDeleter final
+		{
+			void operator()(DynLib* lib) const
+			{
+				lib->unload();
+				bs_delete<DynLib, GenAlloc>(lib);
+			}
+		};
+
+		using DynLibPtr_t = UPtr<DynLib, GenAlloc, DynLibDeleter>;
+
+		struct DynLibPtrComparator final
```

Comment:
I'd prefer we keep the Map as it was. It's a lot of extra code, and in the header file as well - to me it just looks a lot uglier without much upsides.

Reply:
I think there are performance benefits to converting to "Set" instead of "Map".

Two somewhat distinct things.

1) Set allows us to avoid the allocate-and-copy of an additional String. If we had a read-only-reference-counted string, i'd say that it was a meaningless distinction between the two, and wouldn't care. But allocations and copies are expensive in general (though, how expensive is subjective and context sensitive).

2) If we convert to Set, but otherwise don't care about switching to using the "lower_bound" function to save on comparisons when iterating over the Map/Set, then we can use a comparitor that only has the single operator(). However, in order to support the use of the lower_bound function, which I'm almost certain has better performance when needing to insert new values (because it skips a lot of iterating on the insert_hint call), we need to provide a transparent comparitor so that we can search directly for the string, without having allocated a DynLib* already.

I suppose it would be possible to make a rough implementation of lower_bound directly in the cpp file (where lower_bound is called now), and skip the need for the extra two operator() functions. Would you rather do that?

Reply:
I'm aware this was done for performance reasons, but it's highly unlikely DynLib is ever going to be on a performance critical path. When it's not performance critical I value cleaner/smaller code and design in general.

But if you can move the implementation in the .cpp file, in any form, I'd be happy with that.

Reply:
That's very reasonable.

Let me clarify my behavior / motivation here a little bit. I'm mostly a "Can't see the forest for the trees" kind of person, and I have a difficult time understanding how to use something if I haven't taken it apart and understood the relationships between all the internal parts.

So what I've been doing is going through the foundational code for bsf, file by file, and making sure I understand how it functions piece by piece.

So all of these PRs that I've been sending in aren't intended as any kind of criticism or anything. I'm just trying to understand how everything fits together and I figured I might as well try to feed some (relatively minor) improvements back to you while I do it.

So it's really not a huge deal for me, one way or another, if you don't want to accept a particular PR. The act of creating the PR was itself the goal, so that I could understand the code.

That being said, sure, I'll try to move more of the logic into the cpp file. I agree with your stance on trying to keep header files simple and small.